net/dns/resolved.go

```diff
@@ -74,73 +76,164 @@ type resolvedLinkDomain struct {
-		cancelSyncer: cancel,
-		syncerDone:   make(chan struct{}),
-		resolved:     conn.Object(dbusResolvedObject, dbus.ObjectPath(dbusResolvedPath)),
-	}
-	signals := make(chan *dbus.Signal, 16)
-	go ret.resync(ctx, signals)
-	// Only receive the DBus signals we need to resync our config on
-	// resolved restart. Failure to set filters isn't a fatal error,
-	// we'll just receive all broadcast signals and have to ignore
-	// them on our end.
-	if err := conn.AddMatchSignal(dbus.WithMatchObjectPath(dbusPath), dbus.WithMatchInterface(dbusInterface), dbus.WithMatchMember(dbusOwnerSignal), dbus.WithMatchArg(0, dbusResolvedObject)); err != nil {
-		logf("[v1] Setting DBus signal filter failed: %v", err)
+	mgr := &resolvedManager{
+		ctx:    ctx,
+		cancel: cancel,
+
+		logf:  logf,
+		ifidx: iface.Index,
+
+		configCR: make(chan changeRequest),
 	}
-	conn.Signal(signals)
-	return ret, nil
+
+	go mgr.run(ctx)
+
+	return mgr, nil
 }
 
 func (m *resolvedManager) SetDNS(config OSConfig) error {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+	errc := make(chan error, 1)
+	defer close(errc)
+
+	select {
+	case <-m.ctx.Done():
+		return m.ctx.Err()
+	case m.configCR <- changeRequest{config, errc}:
 
-	m.config = config
-	return m.syncLocked(context.TODO()) // would be nice to plumb context through from SetDNS
+	}
+
+	select {
+	case <-m.ctx.Done():
+		return m.ctx.Err()
+	case err := <-errc:
+		m.logf("Resolved systemBus connection interrupted with error: %v.", err)
+		return err
+	}
 }
 
-func (m *resolvedManager) resync(ctx context.Context, signals chan *dbus.Signal) {
-	defer close(m.syncerDone)
+func (m *resolvedManager) run(ctx context.Context) {
+	var (
+		conn      *dbus.Conn
+		signals   chan *dbus.Signal
+		r1Manager dbus.BusObject
+	)
+	bo := backoff.NewBackoff("resolvedSetDNS", m.logf, 30*time.Second)
+	needsReconnect := make(chan bool, 1)
+	defer close(needsReconnect)
+	defer close(m.configCR)
+	defer func() {
+		if conn != nil {
+			conn.Close()
+		}
+	}()
+
+	// reconnect the systemBus if disconnected.
+	reconnect := func() error {
+		var err error
+		signals = make(chan *dbus.Signal, 16)
+		conn, err = dbus.SystemBus()
+		m.logf("[v1] Connecting DBUS or reconnecting to recover disconnect.")
+
+		if err != nil {
+			// Backoff increases time between reconnect attempts.
+			go func() {
+				bo.BackOff(ctx, err)
+				needsReconnect <- true
+			}()
+			return err
+		}
+
+		r1Manager = conn.Object(dbusResolvedObject, dbus.ObjectPath(dbusResolvedPath))
+
+		// Only receive the DBus signals we need to resync our config on
+		// resolved restart. Failure to set filters isn't a fatal error,
+		// we'll just receive all broadcast signals and have to ignore
+		// them on our end.
+		if err = conn.AddMatchSignal(dbus.WithMatchObjectPath(dbusPath), dbus.WithMatchInterface(dbusInterface), dbus.WithMatchMember(dbusOwnerSignal), dbus.WithMatchArg(0, dbusResolvedObject)); err != nil {
+			m.logf("[v1] Setting DBus signal filter failed: %v", err)
+		}
```

Comment:
this error is fine to ignore?

Reply:
Original code ignored it and its noted as not being a fatal error because the code has more broadcast signals to review.